### ISSUE TITLE
xargs: cat: terminated by signal 13 when building library with kraken2

### ISSUE THREAD

**Issue by Veroman** [DerrickWood/kraken2#58]
Hi,
I tried to run the command kraken2-build  --build but this is the output I get:

Creating sequence ID to taxonomy ID map (step 1)...
Sequence ID to taxonomy ID map complete. [0.052s]
Estimating required capacity (step 2)...
xargs: cat: terminated by signal 13

I tried to change the threads, increase the memory, change the k and l values but I still have the same problem. 
Could you please tell me what could be the problem.
thanks!

**Comment by agofton** [comment]
I also have this same error, which occurs regardless of which db I am attempting to build or how much mem/CPUs I give kraken2. I'm almost certain this is an issue from xargs and not from k2 itself. Issue presents at line 125 of `build_kraken2_db.sh`, when xargs tries to parse library.fna/library.faa to build_db. I believe signal 13 means that `build_db` is receiving the input from `xargs -0 cat` and closing the pipe before `xargs -0` cat has finished its job, causing a broken pipe and therefor an error. A clumsy fix to this is to edit `build_kraken2_db.sh` to directly specify library.fna to build_bd like this:

`build_db -k $KRAKEN2_KMER_LEN -l $KRAKEN2_MINIMIZER_LEN -S $KRAKEN2_SEED_TEMPLATE $KRAKEN2XFLAG \
             -H hash.k2d.tmp -t taxo.k2d.tmp -o opts.k2d.tmp -n taxonomy/ -m $seqid2taxid_map_file \
-c $required_capacity -p $KRAKEN2_THREAD_CT $max_db_flag < /path/to/library.fna`

This worked for me but the `/path/to/library.fna` may need to be changed for each db you build and if you have multiple fasta files that you have joined together into a db then you will need to cat the individual library.fna from each of these together and then specify the concatenated file. @DerrickWood @Veroman #

**Comment by Veroman** [comment]
Thanks for the suggestions. I tried to modify that part of build_kraken2_db.sh but I still get the same error. I also tried  to change it at the part of: 

function list_sequence_files() {
  find /path/to/library.fna -print0

but I still get the errors. I think it might kraken
@agofton @DerrickWood #

**Comment by kopardev** [comment]
Any solution to this? I am getting the same error.

**Comment by Grumium** [comment]
This is still unresolved. I cannot even build the standard NCBI plant database. 
The mentioned fix is not working. In my case I could verify that the path to the library is correctly transferred, but the script fails somewhere inside `build_db.cc`. It never reaches the last if clause in the subsequent code snippet, but fails after a random number of processed sequences:

```
    while (true) {
      // Declaration of "ok" and break need to be done outside of critical
      // section to conform with OpenMP spec.
      bool ok;
      #pragma omp critical(reader)
      ok = reader.LoadBlock(std::cin, opts.block_size);
      if (! ok)
        break;
      while (reader.NextSequence(sequence)) {
        auto all_sequence_ids = ExtractNCBISequenceIDs(sequence.header);
        uint64_t taxid = 0;
        for (auto &seqid : all_sequence_ids) {
          auto ext_taxid = ID_to_taxon_map[seqid];
          taxid = taxonomy.LowestCommonAncestor(taxid, taxonomy.GetInternalID(ext_taxid));
        }
        if (taxid) {
          // Add terminator for protein sequences if not already there
          if (opts.input_is_protein && sequence.seq.back() != '*')
            sequence.seq.push_back('*');
          ProcessSequence(sequence.seq, taxid, kraken_index, taxonomy, scanner,
            opts.min_clear_hash_value);
          #pragma omp atomic
          processed_seq_ct++;
          #pragma omp atomic
          processed_ch_ct += sequence.seq.size();
        }
      }
      if (isatty(fileno(stderr))) {
        #pragma omp critical(status_update)
        std::cerr << "\rProcessed " << processed_seq_ct << " sequences (" << processed_ch_ct << " " << (opts.input_is_protein ? "aa" : "bp") << ")...";
      }
    }
  }
  if (isatty(fileno(stderr)))
    std::cerr << "\r";
  std::cerr << "Completed processing of " << processed_seq_ct << " sequences, " << processed_ch_ct << " " << (opts.input_is_protein ? "aa" : "bp") << std::endl;
}
```

**Comment by ddb-repo** [comment]
Did you try running with 1 thread?  That seemed to get me past the signal 13 error but is not ideal.

**Comment by yifengyuan** [comment]
@jenniferlu717 

Dear Jenny,

Could you advice please? I am stuck with the same error. Thank you.

**Comment by yifengyuan** [comment]
With command:
sbatch -p sched_mem1TB -c 40 -t 5-00:00:00 --mem=500000 -J Krakenbulidlib
kraken2-build --build --db myDB

I have trouble with:
Creating sequence ID to taxonomy ID map (step 1)...
Sequence ID to taxonomy ID map already present, skipping map creation.
Estimating required capacity (step 2)...
Estimated hash table requirement: 278027802328 bytes
Capacity estimation complete. [6h34m36.821s]
Building database files (step 3)...
build_db: attempt to create taxonomy w/ node ID == 0
xargs: cat: terminated by signal 13

Thank you!

**Comment by yifengyuan** [comment]
I probably solved it by editing build_kraken2_db.sh at line 125 and 128 as below. The idea is to use the ‘for’ command which would ignore signal 13 instead of the 'xargs' command. I appreciate your comments. Thank you.

**for file in $(find library/ -name '*.fna') ; do cat $file |\\**
build_db -k $KRAKEN2_KMER_LEN -l $KRAKEN2_MINIMIZER_LEN -S $KRAKEN2_SEED_TEMPLATE $KRAKEN2XFLAG \
-H hash.k2d.tmp -t taxo.k2d.tmp -o opts.k2d.tmp -n taxonomy/ -m $seqid2taxid_map_file \
-c $required_capacity -p $KRAKEN2_THREAD_CT $max_db_flag **; done**

**Comment by jkin2003** [comment]
I am getting the same error

**Comment by yifengyuan** [comment]
> I am getting the same error

Did you add sequences to make a custom library? After re-format the local fasta sequences and nodes.dmp, names.dmp, seqid2taxid.dmp, it works well with the latest version of Kraken2.

**Comment by jkin2003** [comment]
I downloaded and installed the bacterial reference library using: 
kraken2-build --download-library bacteria --db $DBNAME

After this process completed I attempted to build the database with:
kraken2-build --build --db $DBNAME
and got the error: 
Estimating required capacity (step 2)...
/work/k/kinskeyj/KrakenDownload/kraken2/scripts/build_kraken2_db.sh: line 101: estimate_capacity: command not found
xargs: cat: terminated by signal 13

**Comment by jkin2003** [comment]
same error except with standard db build

in:
./kraken2-build  --standard  --use-ftp --db ./standardDB

out: 
Step 1/2: Performing ftp file transfer of requested files
Step 2/2: Assigning taxonomic IDs to sequences
All files processed, cleaning up extra sequence files... done, library complete.
Masking low-complexity regions of downloaded library... done.
Step 1/2: Performing ftp file transfer of requested files
Step 2/2: Assigning taxonomic IDs to sequences
All files processed, cleaning up extra sequence files... done, library complete.
Masking low-complexity regions of downloaded library... done.
Step 1/2: Performing ftp file transfer of requested files
Step 2/2: Assigning taxonomic IDs to sequences
All files processed, cleaning up extra sequence files... done, library complete.
Masking low-complexity regions of downloaded library... done.
Step 1/2: Performing ftp file transfer of requested files
Step 2/2: Assigning taxonomic IDs to sequences
All files processed, cleaning up extra sequence files... done, library complete.
Downloading UniVec_Core data from server... done.
Adding taxonomy ID of 28384 to all sequences... done.
Masking low-complexity regions of downloaded library... done.
Creating sequence ID to taxonomy ID map (step 1)...
Sequence ID to taxonomy ID map already present, skipping map creation.
Estimating required capacity (step 2)...
/work/k/kinskeyj/KrakenDownload/kraken2/scripts/build_kraken2_db.sh: line 101: estimate_capacity: command not found
xargs: cat: terminated by signal 13

**Comment by jenniferlu717** [comment]
It seems like the various issues listed in this thread are due to multiple different problems. 
In order for us to help debug, I'm going to close this issue and ask that you open a new issue with 
1) the version of kraken2 you are using (the newest version is suggested) 
2) the command lines used to install kraken (please run `sh install_kraken2.sh` before trying to run any commands) 
3) the command lines used to download the database
3) the command lines used to build the database.

@jkin2003  I believe you need to run `sh install_kraken2.sh ` in order for the `estimate_capacity` command to be found. 

Thank you.

**Comment by jkin2003** [comment]
Hello Jennifer,

Yes the problem was that the kraken2-build script could not find the estimate capacity script, but it was because it could not find the estimate_capacity script in the directory it was installed to. (note: I installed kraken2 and did not move any files following install).  

I added the following line to my sbatch and successfully installed with the  kraken2-build script and that remedied the problem:

export PATH="~/kraken2/src:$PATH"

**Comment by skDooley** [comment]
I am getting the same issue. originally I installed kraken with conda from bioconda. I am running version 2.1.0. I tried doing what jkin2003 tried, but that did not work. I modified the script to echo what the command is and this is the result:

build_db -k 35 -l 31 -S 1111111111111111101010101010101  -H hash.k2d.tmp -t taxo.k2d.tmp -o opts.k2d.tmp -n taxonomy/ -m seqid2taxid.map -c 5142308 -p 15  -B 16384 -b 1093 -r 0

I don't know if that tells you anything more, but this is still an issue.
